### ISSUE TITLE
Advanced functionality commands failing on custom emojis/pings

### ISSUE THREAD

**Issue by HorizonShadow** [meew0/discordrb#501]
If I have a command with arguments, if any of the arguments are a user ping, or a custom emoji, the command will simply not run. 

For example, 

```ruby
  add_reaction = ->(event, regex, output) do
    reaction = UserReaction.first(regex: regex)
    if reaction
      'Reaction already exists'
    else
      UserReaction.create(
        regex: regex,
        output: output,
        created_at: Time.now,
        creator: event.author.username
      )
      'Reaction created'
    end
  end
  command [:ar, :add_reaction],
          min_args: 2,
          max_args: 2,
          usage: 'add_reaction [regex] [message]',
          description: 'Create a reaction',
          &add_reaction
```

With an input of `ar test :some_custom_emoji:`, there will be no error, no output, the command will never be run at all. There's a similar command that accepts a user mention rather than an emoji that has the same behavior.

If I use the actual string for the custom emoji, (for example `<:towc:441574097843912725>`), I get the following error:

```ruby
[ERROR : ct-8 @ 2018-05-25 20:00:27.179] Exception: #<TypeError: no implicit conversion of nil into String>
[ERROR : ct-8 @ 2018-05-25 20:00:27.180] /home/josh/.rvm/rubies/ruby-head/lib/ruby/gems/2.5.0/bundler/gems/discordrb-a7baf7b47c05/lib/discordrb/commands/parser.rb:233:in `gsub'
[ERROR : ct-8 @ 2018-05-25 20:00:27.181] /home/josh/.rvm/rubies/ruby-head/lib/ruby/gems/2.5.0/bundler/gems/discordrb-a7baf7b47c05/lib/discordrb/commands/parser.rb:233:in `block in execute_bare'
[ERROR : ct-8 @ 2018-05-25 20:00:27.182] /home/josh/.rvm/rubies/ruby-head/lib/ruby/gems/2.5.0/bundler/gems/discordrb-a7baf7b47c05/lib/discordrb/commands/parser.rb:218:in `each'
[ERROR : ct-8 @ 2018-05-25 20:00:27.182] /home/josh/.rvm/rubies/ruby-head/lib/ruby/gems/2.5.0/bundler/gems/discordrb-a7baf7b47c05/lib/discordrb/commands/parser.rb:218:in `execute_bare'
[ERROR : ct-8 @ 2018-05-25 20:00:27.182] /home/josh/.rvm/rubies/ruby-head/lib/ruby/gems/2.5.0/bundler/gems/discordrb-a7baf7b47c05/lib/discordrb/commands/parser.rb:258:in `execute'
[ERROR : ct-8 @ 2018-05-25 20:00:27.182] /home/josh/.rvm/rubies/ruby-head/lib/ruby/gems/2.5.0/bundler/gems/discordrb-a7baf7b47c05/lib/discordrb/commands/command_bot.rb:443:in `block in execute_chain'
```

This works with advanced_functionality set to false.

**Comment by z64** [comment]
Could you please share your `CommandBot.new` line?

**Comment by HorizonShadow** [comment]
`BOT = Discordrb::Commands::CommandBot.new token: CONFIG['discord'], prefix: '!!', advanced_functionality: true`

**Comment by z64** [comment]
I can't reproduce this exception with the details given. You should note however that emoji markdown contains default `advanced_functionality` characters that will adversely affect parsing:

The exception comes from here:
```rb
231-        # Append a previous sign if none is present
232-        arguments += @attributes[:previous] unless arguments.include? @attributes[:previous]
233:        arguments = arguments.gsub @attributes[:previous], prev
```
And the defaults are laid out as:
```rb
110-        # All of the following need to be one character
111:        # String to designate previous result in command chain
112:        previous: attributes[:previous] || '~',
113-
114-        # Command chain delimiter
115-        chain_delimiter: attributes[:chain_delimiter] || '>',
116-
117-        # Chain argument delimiter
118-        chain_args_delim: attributes[:chain_args_delim] || ':',
119-
120-        # Sub-chain starting character
121-        sub_chain_start: attributes[:sub_chain_start] || '[',
122-
123-        # Sub-chain ending character
124-        sub_chain_end: attributes[:sub_chain_end] || ']',
125-
126-        # Quoted mode starting character
127-        quote_start: attributes[:quote_start] || '"',
128-
129-        # Quoted mode ending character
130-        quote_end: attributes[:quote_end] || attributes[:quote_start] || '"',
```
You can see that the default values for `chain_delimiter` and `chain_args_delim` will create issues with emoji markdown. You can try adjusting these, but I recommend just not using `advanced_functionality` unless you're actually going to use it. Someone might suggest more "sensible" defaults that won't conflict with emoji markdown, but I don't know what that should be (PRs welcome)

**Comment by z64** [comment]
Okay, I could reproduce this by:

`!!test <:rubytaco:315242245274075157> foo`

Some additional logging:
```
[DEBUG : websocket @ 2018-05-26 01:46:02.984] Raised a Discordrb::Events::MessageEvent
[DEBUG : ct-7 @ 2018-05-26 01:46:02.986] Parsing command chain test <:rubytaco:315242245274075157> foo
[DEBUG : ct-7 @ 2018-05-26 01:46:02.987] Executing bare chain
[DEBUG : ct-7 @ 2018-05-26 01:46:02.987] Executing command rubytaco:315242245274075157 with arguments []
[ERROR : ct-7 @ 2018-05-26 01:46:02.987] Exception: #<TypeError: no implicit conversion of nil into String>
[ERROR : ct-7 @ 2018-05-26 01:46:02.988] /home/lune/git/discordrb/lib/discordrb/commands/parser.rb:233:in `gsub'
[ERROR : ct-7 @ 2018-05-26 01:46:02.988] /home/lune/git/discordrb/lib/discordrb/commands/parser.rb:233:in `block in execute_bare'
[ERROR : ct-7 @ 2018-05-26 01:46:02.988] /home/lune/git/discordrb/lib/discordrb/commands/parser.rb:218:in `each'
[ERROR : ct-7 @ 2018-05-26 01:46:02.988] /home/lune/git/discordrb/lib/discordrb/commands/parser.rb:218:in `execute_bare'
[ERROR : ct-7 @ 2018-05-26 01:46:02.988] /home/lune/git/discordrb/lib/discordrb/commands/parser.rb:258:in `execute'
[ERROR : ct-7 @ 2018-05-26 01:46:02.989] /home/lune/git/discordrb/lib/discordrb/commands/command_bot.rb:443:in `block in execute_chain'
```
I'll try to look into fixing this - it looks like it just fails because it doesn't stop when a command doesn't exist in the middle of a chain..? I'm not looking too closely at this at the moment.. Note that it still probably won't work for you unless you further configure the advanced parser.

**Comment by HorizonShadow** [comment]
I was using advanced functionality because parsing arguments encapsulated by quotes isn't part of the standard bot. 

So one "two three" is 3 arguments without advanced functionality turned out.

I'll check out the parsing options, might be able to make it work with that.

**Comment by z64** [comment]
The exception from this issue has been fixed with #501 (thanks again @unleashy !).

I'm still unsure what acceptable defaults for these options would be so that a command like this parses well for the common use case without breaking things; anyone is free to offer suggestions/patches if this issue has impacted you significantly as well.
